src/node/oscillator.rs

```diff
@@ -0,0 +1,1006 @@
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    ///    use web_audio_api::context::{AudioContext, AsBaseAudioContext};
+    ///    use web_audio_api::node::{PeriodicWave, PeriodicWaveOptions};
+    ///
+    ///    let context = AudioContext::new();
+    ///
+    ///    let options = PeriodicWaveOptions {
+    ///    real: vec![0.,1.,1.],
+    ///    imag: vec![0.,0.,0.],
+    ///    disable_normalization: false,
+    ///    };
+    ///
+    ///    let periodic_wave = PeriodicWave::new(&context, Some(options));
+    /// ```
+    ///
+    pub fn new<C: AsBaseAudioContext>(_context: &C, options: Option<PeriodicWaveOptions>) -> Self {
+        if let Some(PeriodicWaveOptions {
+            real,
+            imag,
+            disable_normalization,
+        }) = options
+        {
+            // Todo: assertions not complete. Missing assertions when
+            // only real field is specified, anly imag field is specified,
+            // and neither real or imag is specified
+            assert!(
+                real.len() >= 2,
+                "RangeError: Real field length should be at least 2"
+            );
+            assert!(
+                imag.len() >= 2,
+                "RangeError: Imag field length should be at least 2",
+            );
+            assert!(
+                real.len() == imag.len(),
+                "RangeError: Imag and real field length should be equal"
+            );
+            Self {
+                real,
+                imag,
+                disable_normalization,
+            }
+        } else {
+            Self {
+                real: vec![0., 0.],
+                imag: vec![0., 1.],
+                disable_normalization: false,
+            }
+        }
+    }
+}
+
+/// Options for constructing an OscillatorNode
+#[derive(Debug)]
+pub struct OscillatorOptions {
+    pub type_: OscillatorType,
+    pub frequency: f32,
+    pub detune: f32,
+    pub channel_config: ChannelConfigOptions,
+    pub periodic_wave: Option<PeriodicWave>,
+}
+
+impl Default for OscillatorOptions {
+    fn default() -> Self {
+        Self {
+            type_: OscillatorType::default(),
+            frequency: 440.,
+            detune: 0.,
+            channel_config: ChannelConfigOptions::default(),
+            periodic_wave: None,
+        }
+    }
+}
+
+/// Waveform of an oscillator
+#[derive(Debug, Copy, Clone)]
+pub enum OscillatorType {
+    Sine,
+    Square,
+    Sawtooth,
+    Triangle,
+    Custom,
+}
+
+impl Default for OscillatorType {
+    fn default() -> Self {
+        OscillatorType::Sine
+    }
+}
+
+impl From<u32> for OscillatorType {
+    fn from(i: u32) -> Self {
+        use OscillatorType::*;
+
+        match i {
+            0 => Sine,
+            1 => Square,
+            2 => Sawtooth,
+            3 => Triangle,
+            4 => Custom,
+            _ => unreachable!(),
+        }
+    }
+}
+
+enum OscMsg {
```

Comment:
What's the motivation for a single discriminant enum?

Reply:
In case i need to add other type of messages.

I had planned to make parameters computations in the control thread and pass them to render thread. like you suggest it in another comment

> Ideally, these computations (and allocations) could be done on the control thread.
> You could extend the OscMsg to include the pre-calculated fields

and i see reading it that you thought about it too. We are on the same page.